pkg/grpc/pool.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpc
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	clocklib "github.com/benbjohnson/clock"
+	"google.golang.org/grpc"
+
+	"github.com/dapr/kit/ptr"
+)
+
+// Real-time clock (wrapper around time.Time) to allow mocking
+var clock = clocklib.New()
+
+// Maximum number of concurrent streams in a single gRPC connection
+// This is the default value used by gRPC servers and clients
+const grpcMaxConcurrentStreams = 100
+
+// ConnectionPool holds a pool of connections to the same address.
+type ConnectionPool struct {
```

Comment:
Is there a way for us to use a library to do pooling instead of writing one ourselves?

Reply:
Not sure if there's any optimized for our specific use case.